app/src/org/commcare/dalvik/odk/provider/InstanceProvider.java

```diff
@@ -365,10 +386,95 @@ public int update(Uri uri, ContentValues values, String where, String[] whereArg
         }
 
+        // If we've changed a particular form instance's status, and not
+        // created a new entry (hence count > 0 check), we need to mirror the
+        // change in that form's record.  NOTE: this conditional is crucial,
+        // since updating a form record in turn calls this update function, so
+        // we need to break the infinite loop by only updating the form record
+        // when the the status changes.
+        if (values.containsKey(InstanceColumns.STATUS) && count > 0) {
+            try {
+                linkToSessionFormRecord(uri);
+            } catch (Exception e) {
+                // TODO: correctly exit function with error code so that
+                // callers can propogate the failure
+            }
+        }
+
         getContext().getContentResolver().notifyChange(uri, null);
+
         return count;
     }
 
+    /**
+     * Register instance with the session's form record.
+     *
+     * @param instanceUri points to a concrete instance we want to register
+     */
+    private void linkToSessionFormRecord(Uri instanceUri) {
+        AndroidSessionWrapper currentState = CommCareApplication._().getCurrentSessionWrapper();
+
+        if (instanceUri == null) {
+            raiseFormEntryError("Form Entry did not return a form", currentState);
+            return;
+        }
+
+        Cursor c = getContext().getContentResolver().query(instanceUri, null, null, null, null);
+        boolean complete = false;
+        try {
+            // register the instance uri and its status with the session
+            // NOTE: there is no safety checks preventing us from binding two
+            // different form instances to the form record found in the
+            // session. We just need to be careful to flush the session's form
+            // record before modifying a new form instance.
+            complete = currentState.beginRecordTransaction(instanceUri, c);
+        } catch (IllegalArgumentException iae) {
+            iae.printStackTrace();
+            // TODO: Fail more hardcore here? Wipe the form record and its ties?
+            raiseFormEntryError("Unrecoverable error when trying to read form|" + iae.getMessage(),
+                    currentState);
+            return;
+        } finally {
+            c.close();
+        }
+
+        FormRecord current;
+        try {
+            current = currentState.commitRecordTransaction();
+        } catch (Exception e) {
+            // Something went wrong with all of the connections which should exist.
+            FormRecordCleanupTask.wipeRecord(getContext(), currentState);
+
+            // Notify the server of this problem (since we aren't going to crash)
+            ExceptionReportTask ert = new ExceptionReportTask();
+            ert.execute(e);
+
+            raiseFormEntryError("An error occurred: " + e.getMessage() + " and your data could not be saved.", currentState);
```

Comment:
What is the user facing outcome here if, for instance, the form is being saved as incomplete and an error occurs. Does the app let you finish filling out the form, but then it errors when you return because the session is gone?